### PR TITLE
DrivAerML: champion recovery seed=1024 eval-interval=10 (Bucket A)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DM champion config (EMA=0.9995 + gc=0.5, 4L/512d) achieves 3.833% val / 4.685% batch-limited test (full-eval TEST best: 4.324% from run zx5syby1). Stochastic initialization effects are significant at 400 training samples — different seeds explore different loss basins. This run covers seed=1024 to expand our seed ensemble, using eval-interval=10 for dense checkpoint coverage and --no-compile-model to avoid the torch.compile + FixedEMA bug (frozen val metrics).

**CRITICAL BUG NOTE:** Always use `--no-compile-model` with EMA on DM. torch.compile caches parameter data — in-place EMA `param.data.copy_()` is NOT reflected in the compiled forward pass, freezing val metrics at initialization values.

## Instructions

Run the exact DM champion config with seed=1024, dense eval interval, and 600-min budget:

```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --ema-mode fixed \
  --no-compile-model \
  --save-checkpoint \
  --eval-interval 10 \
  --seed 1024 \
  --wandb-group dm-champion-recovery \
  --kill-if-best-val-above 150:8.0 \
  --kill-if-no-improvement 200:0.05
```

**After training completes**, run full-eval on the best checkpoint to get authoritative TEST metric:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --ema-mode fixed \
  --no-compile-model \
  --load-checkpoint <path-to-best-checkpoint> \
  --eval-only
```

## Baseline

- **DM best val:** 3.833% at ep511 (PR #3072, eren, W&B: ncl1dh88)
- **DM best batch-limited TEST:** 4.685%
- **DM authoritative full-eval TEST:** 4.324% (faye run zx5syby1, full-eval on ep517 ckpt from w4ufs8m1)
- **External SOTA target (paper):** <3.71% (AB-UPT, ~500 epochs)
- **Champion config:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995`

## Expected output

Report in PR comments:
1. Best val `surface_rel_l2_pct` and epoch
2. W&B run ID and group: `dm-champion-recovery`
3. Checkpoint path
4. Full-eval TEST `surface_rel_l2_pct` (from --eval-only run)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT franky:

## Results

### Training Run — DM champion config with seed=1024

| Metric | Value |
|--------|-------|
| **Best val `surface_rel_l2_pct`** | **4.103%** (epoch 640) |
| **Baseline val (PR #3072)** | 3.833% (epoch 511) |
| **Baseline batch-limited TEST** | 4.685% |
| **Baseline full-eval TEST** | 4.324% |
| **External SOTA (AB-UPT)** | <3.71% |
| W&B run ID | `nudhbnl2` |
| W&B group | `dm-champion-recovery` |
| Epochs trained | 640 (killed: no improvement >0.05% for 200 epochs) |
| Kill reason | `no significant improvement (>0.05) for 200 epochs (patience=200)` |

### Command used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --ema-mode fixed --no-compile-model \
  --save-checkpoint --eval-interval 10 --seed 1024 \
  --wandb-group dm-champion-recovery \
  --kill-if-best-val-above 150:8.0 --kill-if-no-improvement 200:0.05 \
  --agent franky --wandb-name "franky/dm-seed1024-champion-recovery"
```

### Full-eval TEST — NOT AVAILABLE

**Bug: No checkpoint was saved.** The `--kill-if-no-improvement` condition raises a `RuntimeError` at line 1847 of `train.py`, which exits the training loop before reaching the checkpoint-saving code at line 2008. The `--save-checkpoint` flag is set, but the kill exception bypasses it entirely. Without a checkpoint, `--eval-only` cannot be run.

This means we cannot provide the authoritative full-eval TEST metric for this seed.

### What happened

- Training converged well, reaching single-digit val% by epoch 60 and sub-5% by epoch 190.
- Best val of **4.103%** at epoch 640, which is ~0.27% above the baseline val of 3.833%.
- The model plateaued around epoch 440 (val 4.11%) with only marginal improvements afterward.
- Training was killed after 200 epochs without significant (>0.05%) improvement.
- Seed=1024 reaches a competitive loss basin but slightly worse than the champion seed (3.833%).
- The convergence curve was smooth with occasional loss spikes from hard training samples, but these recovered quickly.
- Grad clip events decreased steadily from ~394/epoch early on to near-zero by epoch 250+, indicating model stability.

### Suggested follow-ups

1. **Fix checkpoint save on kill:** The kill condition should save the best checkpoint before raising the RuntimeError — this is a bug that affects all seed-recovery experiments using `--kill-if-no-improvement`.
2. **Seed ensemble analysis:** Compare val convergence curves across seeds (1024, 456, 777, champion) to understand seed sensitivity at 400 training samples.
3. **Re-run with checkpoint fix:** Once the bug is fixed, re-run to get the authoritative full-eval TEST metric.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3311 because ADVISOR: Closing — seed=1024 MSE-only champion val=4.103% at ep640 (killed by patience). Doesn't beat MSE baseline 3.833% or metric-aware baseline 3.700%. The MSE-only champion config shows significant seed sensitivity (seeds 0/42 hit 3.83%, seed=1024 misses at 4.10%). Metric-aware loss is now the primary direction. Franky reassigned..
